src/main/java/mezz/jei/ingredients/IngredientSorter.java

```diff
@@ -80,4 +103,233 @@ public void invalidateCache() {
+		}
+		return 0;
+	};
+
+	private static int getArmorDamageReduce(ItemStack itemStack) {
+		Item item = itemStack.getItem();	
+		if (item instanceof ArmorItem) {
+			ArmorItem armorItem = (ArmorItem) item;				
+			return armorItem.getDamageReduceAmount();
+		}
+		return Integer.MIN_VALUE;
+	};
+
+	private static float getArmorToughness(ItemStack itemStack) {
+		Item item = itemStack.getItem();	
+		if (item instanceof ArmorItem) {
+			ArmorItem armorItem = (ArmorItem) item;				
+			return armorItem.getToughness();
+		}
+		return Float.MIN_VALUE;
+	};
+
+	private static int getArmorDurability(ItemStack itemStack) {
+		if (isArmor(itemStack))
+			return getMaxDamage(itemStack);
+		return 0;
+	}
+
+	private static String getTagForSorting (IIngredientListElementInfo<?> elementInfo) {
+		IIngredientManager ingredientManager = Internal.getIngredientManager();
+		Collection<ResourceLocation> tags = elementInfo.getTagIds(ingredientManager);
+		String bestTag = "";
+		if (tags.size() > 0 ) {
+			int maxTagSize = Integer.MIN_VALUE;
+			//Group things by the most popular tag it has.
+			for (ResourceLocation tag : tags) {			
+				//TODO: make a tag blacklist.
+				if (!tag.toString().equals("itemfilters:check_nbt")) {
+					int thisTagSize = ItemTags.getCollection().getTagByID(tag).getAllElements().size();
+					if (thisTagSize > maxTagSize) {
+						bestTag = tag.getPath();
+						maxTagSize = thisTagSize;
+					}
+				}
+			}
+		}
+		return bestTag;
+	};
+
+	private static boolean hasTag(IIngredientListElementInfo<?> elementInfo){
+		//Sort non-tools after the tools.
+		return !getTagForSorting(elementInfo).isEmpty();
+	};
+
+	private static String getToolClass(ItemStack itemStack)
+    {
+		//I think I should find a way to cache this.
+        if (itemStack == null || itemStack == ItemStack.EMPTY) return "";
+		Item item = itemStack.getItem();
+        Set<ToolType> toolTypeSet = item.getToolTypes(itemStack);
+        
+        Set<String> toolClassSet = new HashSet<String>();
+
+        for (ToolType toolClass: toolTypeSet) {
+            //Swords are not "tools".
+            if (toolClass.getName() != "sword") {
+            	toolClassSet.add(toolClass.getName());
+            }
+        }
+
+        //Minecraft hoes, shears, and fishing rods don't have tool class names.
+        if (toolClassSet.isEmpty()) {
+            if (item instanceof HoeItem) return "hoe";
+            if (item instanceof ShearsItem) return "shears";
+            if (item instanceof FishingRodItem) return "fishingrod";
+            return "";
+        }
+        
+        //Get the only thing.
+        if (toolClassSet.size() == 1)
+            return (String) toolClassSet.toArray()[0];
+        
+        //We have a preferred type to list tools under, primarily the pickaxe for harvest level.
+        String[] prefOrder = {"pickaxe", "axe", "shovel", "hoe", "shears", "wrench"};
+        for (int i = 0; i < prefOrder.length; i++)
```

Comment:
to follow JEI's style, please use `{` and `}` braces wherever possible

Reply:
Ok.  That got copied verbatim from somewhere else I had written it.  I had noticed the JEI style and applied it to the new stuff I did, but didn't hit this bit.